catanatron_core/catanatron/state.py

```diff
@@ -591,9 +599,90 @@ def apply_action(state: State, action: Action):
+    elif action.action_type == ActionType.OFFER_TRADE:
+        state.is_resolving_trade = True
+        state.current_trade = (*action.value, state.current_turn_index)
+
+        # go in seating order; order won't matter because of "acceptees hook"
+        state.current_player_index = next(
+            i for i, c in enumerate(state.colors) if c != action.color
+        )  # cant ask yourself
+        state.current_prompt = ActionPrompt.DECIDE_TRADE
+
+        state.playable_actions = generate_playable_actions(state)
+    elif action.action_type == ActionType.ACCEPT_TRADE:
+        # add yourself to self.acceptees
+        index = state.colors.index(action.color)
+        new_acceptess = list(state.acceptees)
+        new_acceptess[index] = True  # type: ignore
+        state.acceptees = tuple(new_acceptess)
+
+        try:
+            # keep going around table w/o asking yourself or players that have answered
+            state.current_player_index = next(
+                i
+                for i, c in enumerate(state.colors)
+                if c != action.color and i > state.current_player_index
+            )
+            # .is_resolving_trade, .current_trade, .current_prompt, .acceptees stay the same
+        except StopIteration:
+            # by this action, there is at least 1 acceptee, so go to DECIDE_ACCEPTEES
+            # .is_resolving_trade, .current_trade, .acceptees stay the same
+            state.current_player_index = state.current_turn_index
+            state.current_prompt = ActionPrompt.DECIDE_ACCEPTEES
+
+        state.playable_actions = generate_playable_actions(state)
+    elif action.action_type == ActionType.REJECT_TRADE:
+        try:
+            # keep going around table w/o asking yourself or players that have answered
+            state.current_player_index = next(
+                i
+                for i, c in enumerate(state.colors)
+                if c != action.color and i > state.current_player_index
+            )
+            # .is_resolving_trade, .current_trade, .current_prompt, .acceptees stay the same
+        except StopIteration:
+            # if no acceptees at this point, go back to PLAY_TURN
+            if sum(state.acceptees) == 0:
+                reset_trading_state(state)
+
+                state.current_player_index = state.current_turn_index
+                state.current_prompt = ActionPrompt.PLAY_TURN
+            else:
+                # go to offering player with all the answers
+                # .is_resolving_trade, .current_trade, .acceptees stay the same
+                state.current_player_index = state.current_turn_index
+                state.current_prompt = ActionPrompt.DECIDE_ACCEPTEES
+
+        state.playable_actions = generate_playable_actions(state)
+    elif action.action_type == ActionType.CONFIRM_TRADE:
+        # apply trade
+        offering = action.value[:5]
+        asking = action.value[5:10]
+        enemy = action.value[10]
```

Comment:
what does enemy mean here? Also, it's a little weird to me to have these magic numbers here. If you're set on having it be a single tuple, maybe have a helper to help this decomposition?

Reply:
It's the color of the other end of the trade. Changed to `enemy_color` as an improvement. Open to suggestions.

About magic numbers, make sense. Tho I want to favor performance in the codebase and I think this approach (destructuring the object with variables) gives a good balance of readability and performance.